File: frontend/src/lib/api/counselor-info.ts
// SDD-077 상담사 정보 관리 API — 플랫폼/기관 관리자용 조회·수정 클라이언트

import { apiClient } from './client';
import type { QualificationItem, CareerItem } from './counselor';

export interface CounselorInfoDto {
  id: string;
  email: string;
  name: string;
  role: string;
  status: string;
  org_id: string | null;
  org_name: string | null;
  counselor_code: string | null;
  phone: string | null;
  profile_image: string | null;
  bio: string | null;
  gender: string | null;
  birth_date: string | null;
  postal_code: string | null;
  address_line1: string | null;
  address_line2: string | null;
  affiliation_type: string | null;
  years_of_experience: number | null;
  specialties: string[];
  qualifications: QualificationItem[];
  careers: CareerItem[];
  version: number;
}

export interface CounselorInfoUpdatePayload {
  name?: string;
  phone?: string | null;
  profile_image?: string | null;
  bio?: string | null;
  gender?: string | null;
  birth_date?: string | null;
  postal_code?: string | null;
  address_line1?: string | null;
  address_line2?: string | null;
  affiliation_type?: string | null;
  years_of_experience?: number | null;
  specialties?: string[];
  qualifications?: QualificationItem[];
  careers?: CareerItem[];
  version?: number;
  // 관리자 수정 사유 (본인 수정에서는 미사용)
  reason?: string;
}

export interface AdminCounselorListItemDto {
  id: string;
  name: string;
  email: string;
  counselor_code: string | null;
  role: string;
  status: string;
  org_id: string | null;
  org_name: string | null;
}

export interface AdminCounselorListResponseDto {
  items: AdminCounselorListItemDto[];
  total: number;
}

export interface PrimaryAdminProfilePatchPayload {
  name?: string;
  phone?: string | null;
  reason: string;
  expected_user_id?: string;
}

export interface PrimaryAdminProfileDto {
  id: string;
  name: string;
  email: string;
  phone: string | null;
  role: string;
  status: string;
}

// --- 플랫폼 관리자 ---

export const listAdminCounselors = (
  params?: { org_id?: string; q?: string; status?: string },
): Promise<AdminCounselorListResponseDto> => {
  const sp = new URLSearchParams();
  for (const [key, value] of Object.entries(params ?? {})) {
    if (value) sp.set(key, value);
  }
  const qs = sp.toString();
  return apiClient.get<AdminCounselorListResponseDto>(`/admin/counselors${qs ? `?${qs}` : ''}`);
};

export const getAdminCounselorProfile = (userId: string): Promise<CounselorInfoDto> =>
  apiClient.get<CounselorInfoDto>(`/admin/counselors/${userId}/profile`);

export const patchAdminCounselorProfile = (
  userId: string,
  payload: CounselorInfoUpdatePayload,
): Promise<CounselorInfoDto> =>
  apiClient.patch<CounselorInfoDto>(`/admin/counselors/${userId}/profile`, payload);

export const patchPrimaryAdminProfile = (
  orgId: string,
  payload: PrimaryAdminProfilePatchPayload,
): Promise<PrimaryAdminProfileDto> =>
  apiClient.patch<PrimaryAdminProfileDto>(`/admin/orgs/${orgId}/primary-admin/profile`, payload);

// --- 기관 관리자 ---

export const getOrgCounselorProfile = (orgId: string, userId: string): Promise<CounselorInfoDto> =>
  apiClient.get<CounselorInfoDto>(`/org/${orgId}/counselors/${userId}/profile`);

export const patchOrgCounselorProfile = (
  orgId: string,
  userId: string,
  payload: CounselorInfoUpdatePayload,
): Promise<CounselorInfoDto> =>
  apiClient.patch<CounselorInfoDto>(`/org/${orgId}/counselors/${userId}/profile`, payload);

File: backend/app/schemas/counselor_info.py
"""SDD-077 상담사 정보 관리 스키마 — 3자(본인/플랫폼/기관) 공용 조회·수정 계약"""

from pydantic import BaseModel, ConfigDict, Field

from app.schemas.auth import CareerItem, QualificationItem

# 대량 할당 차단 대상 — 전송 자체를 403으로 거부한다 (별도 관리 기능 이용)
FORBIDDEN_UPDATE_FIELDS = ("email", "role", "org_id", "verified_tier", "status", "counselor_code")


class CounselorInfoUpdate(BaseModel):
    """상담사 정보 수정 요청 — 미전송은 유지, null은 삭제(선택 필드만)."""

    model_config = ConfigDict(extra="forbid")

    # 계정 기본 (User)
    name: str | None = Field(None, max_length=100)
    phone: str | None = Field(None, max_length=20)
    profile_image: str | None = Field(None, max_length=500)
    bio: str | None = None
    # 개인정보 (CounselorProfile) — 선택 입력, null 보존
    gender: str | None = Field(None, pattern="^(male|female|other)$")
    birth_date: str | None = Field(None, description="YYYY-MM-DD")
    postal_code: str | None = Field(None, max_length=20)
    address_line1: str | None = Field(None, max_length=300)
    address_line2: str | None = Field(None, max_length=200)
    # 전문 이력
    affiliation_type: str | None = Field(None, max_length=50)
    years_of_experience: int | None = Field(None, ge=0)
    specialties: list[str] | None = None
    qualifications: list[QualificationItem] | None = None
    careers: list[CareerItem] | None = None
    # 낙관적 잠금 (미전송 시 검사 생략 — 구 클라이언트 하위 호환)
    version: int | None = Field(None, ge=1)
    # 관리자 수정 사유 (본인 수정은 불필요)
    reason: str | None = Field(None, max_length=2000)
    # 수정 금지 필드 — 선언만 해두고 전송되면 403으로 거부한다
    email: str | None = None
    role: str | None = None
    org_id: str | None = None
    verified_tier: str | None = None
    status: str | None = None
    counselor_code: str | None = None


class CounselorInfoResponse(BaseModel):
    """상담사 정보 응답 — 이메일은 읽기 전용 표시용."""

    id: str
    email: str
    name: str
    role: str
    status: str
    org_id: str | None = None
    org_name: str | None = None
    counselor_code: str | None = None
    phone: str | None = None
    profile_image: str | None = None
    bio: str | None = None
    gender: str | None = None
    birth_date: str | None = None
    postal_code: str | None = None
    address_line1: str | None = None
    address_line2: str | None = None
    affiliation_type: str | None = None
    years_of_experience: int | None = None
    specialties: list[str] = []
    qualifications: list[QualificationItem] = []
    careers: list[CareerItem] = []
    version: int = 1


class AdminCounselorListItem(BaseModel):
    id: str
    name: str
    email: str
    counselor_code: str | None = None
    role: str
    status: str
    org_id: str | None = None
    org_name: str | None = None


class AdminCounselorListResponse(BaseModel):
    items: list[AdminCounselorListItem]
    total: int


class PrimaryAdminProfilePatch(BaseModel):
    """주 담당자 이름/전화 수정 — 담당자 교체(primary_admin_id 변경) 아님."""

    model_config = ConfigDict(extra="forbid", str_strip_whitespace=True)

    name: str | None = Field(None, min_length=1, max_length=100)
    phone: str | None = Field(None, max_length=20)
    reason: str = Field(min_length=1, max_length=2000)
    # 저장 직전 담당자 교체 충돌 감지용 — 불일치 시 409
    expected_user_id: str | None = None
